File: Referencias/Repos SoyMomo/soymomo-cs-platform-backend-main/soymomo-cs-platform-backend-main/src/models/Watch/Wearer.ts
import type { File, GeoPoint } from 'parse/node';
import Parse from 'parse/node';

import type User from './User';
import WatchSettings from './WatchSettings';

export default class Wearer extends Parse.Object {
  static className = 'Wearer';

  constructor() {
    super('Wearer');
  }

  get deviceId(): string {
    return this.get('deviceId');
  }

  set deviceId(value: string) {
    this.set('deviceId', value);
  }

  get firstName(): string {
    return this.get('firstName');
  }

  set firstName(value: string) {
    this.set('firstName', value);
  }

  get lastName(): string {
    return this.get('lastName');
  }

  set lastName(value: string) {
    this.set('lastName', value);
  }

  get phone(): string {
    return this.get('phone');
  }

  set phone(value: string) {
    this.set('phone', value);
  }

  get imei(): string {
    return this.get('imei');
  }

  set imei(value: string) {
    this.set('imei', value);
  }

  get weight(): number {
    return this.get('weight');
  }

  set weight(value: number) {
    this.set('weight', value);
  }

  get height(): number {
    return this.get('height');
  }

  set height(value: number) {
    this.set('height', value);
  }

  get birthday(): Date | null {
    return this.get('birthday');
  }

  set birthday(value: Date | null) {
    this.set('birthday', value);
  }

  get steps(): number {
    return this.get('steps');
  }

  set steps(value: number) {
    this.set('steps', value);
  }

  get deviceManufacturer(): string {
    return this.get('deviceManufacturer');
  }

  set deviceManufacturer(value: string) {
    this.set('deviceManufacturer', value);
  }

  get userInCharge(): User {
    return this.get('userInCharge');
  }

  set userInCharge(value: User) {
    this.set('userInCharge', value);
  }

  get settings(): WatchSettings {
    return this.get('settings');
  }

  set settings(value: WatchSettings) {
    this.set('settings', value);
  }

  get lastKnownLocation(): GeoPoint {
    return this.get('lastKnownLocation');
  }

  set lastKnownLocation(value: GeoPoint) {
    this.set('lastKnownLocation', value);
  }

  get lastLocationTime(): Date {
    return this.get('lastLocationTime');
  }

  set lastLocationTime(value: Date) {
    this.set('lastLocationTime', value);
  }

  get lastAccuracy(): number {
    return this.get('lastAccuracy');
  }

  set lastAccuracy(value: number) {
    this.set('lastAccuracy', value);
  }

  get accuracy(): number {
    return this.get('accuracy');
  }

  set accuracy(value: number) {
    this.set('accuracy', value);
  }

  get batteryPercentage(): number {
    return this.get('batteryPercentage');
  }

  set batteryPercentage(value: number) {
    this.set('batteryPercentage', value);
  }

  get avatarId(): number {
    return this.get('avatarId');
  }

  set avatarId(value: number) {
    this.set('avatarId', value);
  }

  get active(): boolean {
    return this.get('active');
  }

  set active(value: boolean) {
    this.set('active', value);
  }

  get lastTKQ(): Date | null {
    return this.get('lastTKQ');
  }

  set lastTKQ(value: Date | null) {
    this.set('lastTKQ', value);
  }

  get hearts(): number {
    return this.get('hearts');
  }

  set hearts(value: number) {
    this.set('hearts', value);
  }

  get hasWatchOn(): boolean {
    return this.get('hasWatchOn');
  }

  set hasWatchOn(value: boolean) {
    this.set('hasWatchOn', value);
  }

  get GPSMode(): number {
    return this.get('GPSMode');
  }

  set GPSMode(value: number) {
    this.set('GPSMode', value);
  }

  get lastGpsDate(): Date {
    return this.get('lastGpsDate');
  }

  set lastGpsDate(value: Date) {
    this.set('lastGpsDate', value);
  }

  get updatedAt(): Date {
    return this.get('updatedAt');
  }

  set updatedAt(value: Date) {
    this.set('updatedAt', value);
  }

  get fromGps(): boolean {
    return this.get('fromGps');
  }

  set fromGps(value: boolean) {
    this.set('fromGps', value);
  }

  get oldLocation(): string[] {
    return this.get('oldLocation');
  }

  set oldLocation(value: string[]) {
    this.set('oldLocation', value);
  }

  get image(): File | null {
    return this.get('image');
  }

  set image(value: File | null) {
    this.set('image', value);
  }

  get pushy(): string | null {
    return this.get('pushy');
  }

  set pushy(value: string | null) {
    this.set('pushy', value);
  }

  get updatedLocation(): boolean {
    return this.get('updatedLocation');
  }

  set updatedLocation(value: boolean) {
    this.set('updatedLocation', value);
  }

  get model(): number {
    return this.get('model');
  }

  set model(value: number) {
    this.set('model', value);
  }

  get firstLinked(): Date {
    return this.get('firstLinked');
  }

  set firstLinked(value: Date) {
    this.set('firstLinked', value);
  }

  get batterySaveInUse(): boolean {
    return this.get('batterySaveInUse');
  }

  set batterySaveInUse(value: boolean) {
    this.set('batterySaveInUse', value);
  }

  get hardwareModel(): String {
    return this.get('hardwareModel');
  }

  set hardwareModel(value: String) {
    this.set('hardwareModel', value);
  }

  public static register(): void {
    Parse.Object.registerSubclass(this.className, this);
  }

  fromParseObject(parseObject: Parse.Object): Wearer {
    this.id = parseObject.id;
    this.firstName = parseObject.get('firstName');
    this.lastName = parseObject.get('lastName');
    this.weight = parseObject.get('weight');
    this.height = parseObject.get('height');
    this.birthday = parseObject.get('birthday');
    this.steps = parseObject.get('steps');
    this.deviceManufacturer = parseObject.get('deviceManufacturer');
    this.userInCharge = parseObject.get('userInCharge');
    if (parseObject.get('settings')) {
      this.settings = Object.assign(
        new WatchSettings().fromParseObject(parseObject.get('settings')),
        parseObject.get('settings')
      );
    }
    this.lastKnownLocation = parseObject.get('lastKnownLocation');
    this.lastLocationTime = parseObject.get('lastLocationTime');
    this.lastAccuracy = parseObject.get('lastAccuracy');
    this.accuracy = parseObject.get('accuracy');
    this.batteryPercentage = parseObject.get('batteryPercentage');
    this.avatarId = parseObject.get('avatarId');
    this.active = parseObject.get('active');
    this.lastTKQ = parseObject.get('lastTKQ');
    this.updatedAt = parseObject.get('updatedAt');
    this.hearts = parseObject.get('hearts');
    this.hasWatchOn = parseObject.get('hasWatchOn');
    this.lastGpsDate = parseObject.get('lastGpsDate');
    this.fromGps = parseObject.get('fromGps');
    this.phone = parseObject.get('phone');
    this.deviceId = parseObject.get('deviceId');
    this.imei = parseObject.get('imei');
    this.oldLocation = parseObject.get('oldLocation');
    this.image = parseObject.get('image');
    this.pushy = parseObject.get('pushy');
    this.updatedLocation = parseObject.get('updatedLocation');
    this.model = parseObject.get('model');
    this.firstLinked = parseObject.get('firstLinked');
    this.batterySaveInUse = parseObject.get('batterySaveInUse');
    this.hardwareModel = parseObject.get('hardwareModel');
    return this;
  }
}

Wearer.register();

File: Referencias/Repos SoyMomo/soymomo-watch-android-main (1)/soymomo-watch-android-main/app/src/main/java/com/sosmartlabs/momo/models/Wearer.kt
package com.sosmartlabs.momo.models

import com.parse.*
import com.parse.ktx.delegates.ParseDelegate
import com.sosmartlabs.momo.addfirstwatch.model.remote.MobileNetworkOperator
import com.sosmartlabs.momo.firebase.FirebaseRemoteConfigRepository
import com.sosmartlabs.momo.watchcontacts.model.Contact
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.io.path.Path

@ParseClassName("Wearer")
class Wearer : ParseObject() {
    var deviceId by ParseDelegate<String>(null)
    var firstName by ParseDelegate<String>(null)
    var lastName by ParseDelegate<String>(null)
    var lastKnownLocation by ParseDelegate<ParseGeoPoint?>(null)
    var settings by ParseDelegate<WatchSettings>(null)
    var userInCharge by ParseDelegate<ParseUser>(null)
    var hardwareModel by ParseDelegate<String>(null)
    var batteryPercentage by ParseDelegate<Int>(null)
    var steps by ParseDelegate<Int>(null)
    var hearts by ParseDelegate<Int>(null)
    var GPSMode by ParseDelegate<Int>(null)
    var phone by ParseDelegate<String>(null)
    var imei by ParseDelegate<String>(null)
    var contacts by ParseDelegate<ParseRelation<Contact>>(null)
    var accuracy by ParseDelegate<Int>(null)
    var image by ParseDelegate<ParseFile?>(null)
    var birthday by ParseDelegate<Date>(null)
    var weight by ParseDelegate<Int?>(null)
    var height by ParseDelegate<Int?>(null)
    var watchColor by ParseDelegate<String?>(null)
    var mobileNetworkOperator by ParseDelegate<MobileNetworkOperator?>(null)
    var firstInteractionNotified by ParseDelegate<Boolean>(null)
    var lastLocationTime by ParseDelegate<Date?>(null)

    var profileModified: Boolean = false

    val model by lazy {
        when (modelInt) {
            0 -> Original
            1 -> H2OChile
            2 -> H2OSpain
            3 -> H2OEurope
            4 -> H2OChileAmPm
            5 -> Space1
            6 -> Space2
            7 -> Lite1
            8 -> Space3
            9 -> Space4
            else -> {
                Timber.w("Wearer: $deviceId unknown model int: $modelInt, defaulting to Original")
                Original
            }
        }
    }

    val modelInt: Int
        get() = if (has("model")) getInt("model") else {
            Timber.d("Wearer: $deviceId model not set, defaulting to Space 2.0 (6)")
            6 // Defaults to Space 2.0.
        }

    val showBatteryWarning: Boolean
        get() = if (has("GPSMode")) {
            when (modelInt) {
                0, 1, 2, 3, 4 -> GPSMode in 2..3
                5, 6, 7 -> GPSMode == 4
                8, 9 -> false // Space 3.0 and Space 4.0: skip GPS mode warnings
                else -> {
                    Timber.w("Wearer: $deviceId unknown model int for battery warning: $modelInt")
                    false
                }
            }
        } else {
            Timber.d("Wearer: $deviceId GPSMode not set, battery warning disabled")
            false
        }

    // Last connection state this instance logged. isConnected() runs on every card
    // bind, selector highlight, and DiffUtil content check, so it logs only when
    // the computed state changes instead of on every call.
    private var lastLoggedConnectionState: Boolean? = null

    fun isConnected(): Boolean {
        if (!has("lastTKQ")) {
            if (lastLoggedConnectionState != false) {
                lastLoggedConnectionState = false
                Timber.d("Wearer: $deviceId lastTKQ not set, considering as disconnected")
            }
            return false
        }
        val timeSinceLastTKQInMillis = System.currentTimeMillis() - getDate("lastTKQ")!!.time
        val thresholdLastTKQInMillis = TimeUnit.MINUTES.toMillis(FirebaseRemoteConfigRepository.thresholdMinutesLastTKQ)
        val isConnected = timeSinceLastTKQInMillis <= thresholdLastTKQInMillis
        if (lastLoggedConnectionState != isConnected) {
            lastLoggedConnectionState = isConnected
            Timber.d(
                "Wearer: $deviceId connection status: $isConnected (last TKQ ${TimeUnit.MILLISECONDS.toMinutes(timeSinceLastTKQInMillis)}min ago, threshold ${TimeUnit.MILLISECONDS.toMinutes(thresholdLastTKQInMillis)}min)"
            )
        }
        return isConnected
    }

    /**
     * The last check-in the backend saw, or null if the column is absent.
     *
     * This is the same value [isConnected] thresholds on, so anything shown to the user as
     * "last seen" is derived from the fact that decided they are offline in the first place.
     */
    fun lastCheckInAt(): Date? = if (has("lastTKQ")) getDate("lastTKQ") else null

    /**
     * Last reported charge, or null when the column is absent. Note this is the reading on
     * the way down, not the level now — a disconnected watch has been silent since.
     */
    fun reportedBatteryPercentage(): Int? = if (has("batteryPercentage")) batteryPercentage else null

    /**
     * Whether a disconnect should be attributed to a flat battery rather than lost signal.
     *
     * Read by the disconnected banner, the wearer card and the chat card, so that all three
     * name the same cause for the same watch — they previously disagreed, and one screen
     * could show a flat-battery icon and a no-SIM icon at the same time.
     *
     * An absent reading returns false: with no number on record there is no evidence the
     * battery died, and "it ran out of charge" is not a claim to make on a missing column.
     */
    fun ranOutOfBattery(): Boolean {
        val percentage = reportedBatteryPercentage() ?: return false
        return percentage < DEAD_BATTERY_THRESHOLD_PERCENT
    }

    fun hasVideocall(): Boolean {
        // First we check by model if the videocall capability is supported
        val model = this.model
        val modelSupportsVideoCalls = model.hasVideoCalls()
        Timber.d("Wearer: $deviceId model ${model.javaClass.simpleName} (modelInt: $modelInt) supports video calls: $modelSupportsVideoCalls")
        
        if (!modelSupportsVideoCalls) { 
            Timber.d("Wearer: $deviceId videocall capability: false (model does not support video calls)")
            return false 
        }
        
        // Second we check that the object has the necessary pushy token to receive push notifications
        val hasPushyToken = has("pushy")
        val hasVideocall = hasPushyToken
        Timber.d("Wearer: $deviceId pushy token present: $hasPushyToken")
        Timber.d("Wearer: $deviceId videocall capability: $hasVideocall (model supports: $modelSupportsVideoCalls, pushy token: $hasPushyToken)")
        return hasVideocall
    }

    fun name(): String = when {
        has("firstName") && has("lastName") -> "$firstName $lastName"
        has("firstName") -> firstName
        has("lastName") -> lastName
        has("deviceId") -> deviceId
        else -> {
            Timber.w("Wearer: $deviceId no identifying information found, using objectId")
            objectId
        }
    }

    fun imei(): String = when {
        has("imei") -> imei
        has("deviceId") -> {
            Timber.d("Wearer: $deviceId IMEI not found, using deviceId as fallback")
            deviceId
        }
        else -> {
            Timber.w("Wearer: $deviceId No IMEI or deviceId found, using objectId")
            objectId
        }
    }

    fun deviceId(): String = when {
        has("deviceId") -> deviceId
        has("imei") -> {
            Timber.d("Wearer: $deviceId DeviceId not found, using IMEI as fallback")
            imei
        }
        else -> {
            Timber.w("Wearer: $deviceId No deviceId or IMEI found, using objectId")
            objectId
        }
    }

    fun modelName(): String = when (model) {
        Original -> "SoyMomo Original"
        H2OChile, H2OChileAmPm, H2OEurope, H2OSpain -> "SoyMomo H2O"
        Space1 -> "SoyMomo Space 1.0"
        Space2 -> "SoyMomo Space 2.0"
        Space3 -> "SoyMomo Space 3.0"
        Space4 -> "SoyMomo Space 4.0"
        Lite1 -> "SoyMomo Space Lite 1.0"
        else -> {
            Timber.w("Wearer: $deviceId Unknown model: $model")
            "Unknown model"
        }
    }

    fun shortModelName(): String = when (model) {
        Original -> "Original"
        H2OChile, H2OChileAmPm, H2OEurope, H2OSpain -> "H2O"
        Space1 -> "Space 1.0"
        Space2 -> "Space 2.0"
        Space3 -> "Space 3.0"
        Space4 -> "Space 4.0"
        Lite1 -> "Space Lite 1.0"
        else -> {
            Timber.w("Wearer: $deviceId Unknown model for short name: $model")
            "Unknown model"
        }
    }

    fun hasSuccessfulFirstInteraction(): Boolean {
        if (!has("firstInteractionNotified")) {
            Timber.d("Wearer: $deviceId firstInteraction not set, returning false")
            return false
        }
        Timber.d("Wearer: $deviceId has first interaction: $firstInteractionNotified")
        return firstInteractionNotified
    }

    fun hasUnlimitedMessageLength(): Boolean {
        // Models Space 2.0 (6) and Space 3.0 (8) have unlimited message length
        return model == Space2 || model == Space3 || model == Space4
    }

    fun isSpace4OrNewer(): Boolean {
        return modelInt >= 9
    }

    fun isSpace3OrNewer(): Boolean {
        return modelInt >= 8
    }

    init {
        Timber.d("Wearer: Wearer initialized")
    }

    companion object {
        /**
         * Below this last-reported charge, a disconnect is attributed to the battery having
         * run out; at or above it the charge was healthy when the watch last checked in, so
         * the disconnect is attributed to lost signal.
         *
         * Lives on the model because more than one surface has to agree on it: the
         * disconnected banner on the main map and the wearer card underneath it both pick
         * their glyph and their copy from this split, and when they disagreed the same
         * screen showed a flat-battery icon and a no-SIM icon for one watch.
         */
        const val DEAD_BATTERY_THRESHOLD_PERCENT = 10
    }
}